### ISSUE TITLE
Race condition that leads to a crash?

### ISSUE THREAD

**Issue by groue** [facebook/SocketRocket#332]
Hello,

I write an application which uses web sockets, and has to be very resilient: in case of connection failure, the app reconnects shortly after, and tries to bring back on the service.

In case of server failure, I notice a crash which happens not frequently, but frequently enough to be painful, on https://github.com/square/SocketRocket/blob/0.4.2/SocketRocket/SRWebSocket.m#L1465

To investigate, I added two lines of code which log `-[SRWebSocket dealloc]` (at [line](https://github.com/square/SocketRocket/blob/0.4.2/SocketRocket/SRWebSocket.m#L352)) and `-[SRWebSocket stream:handleEvent:]` (at [line](https://github.com/square/SocketRocket/blob/0.4.2/SocketRocket/SRWebSocket.m#L1446))

```objc
- (void)dealloc
{
+    NSLog(@"%s %p", __PRETTY_FUNCTION__, self);
```

```objc
- (void)stream:(NSStream *)aStream handleEvent:(NSStreamEvent)eventCode
{
    // ...
    dispatch_async(_workQueue, ^{
+        NSLog(@"%s %p", __PRETTY_FUNCTION__, self);
```

And indeed, the last log lines before the crash show that the dispatch block uses a deallocated object `0x195ee290`:

```
2016-02-01 19:49:47.869 MyApp[4377:5674086] -[SRWebSocket dealloc] 0x195ee290
2016-02-01 19:49:47.870 MyApp[4377:5674086] __34-[SRWebSocket stream:handleEvent:]_block_invoke487 0x195ee290
```

What do you think?

**Comment by groue** [comment]
In my specific case, the crash happens when the websocket server is unreachable.

My current workaround is to crash *less often* by waiting longer before reconnecting when this particular kind of error happens:

```objc
#import <netdb.h>  // EAI_NONAME declaration
...
- (void)webSocket:(SRWebSocket *)webSocket didFailWithError:(NSError *)error
{
    if ([error.domain isEqualToString:(NSString *)kCFErrorDomainCFNetwork] &&
        error.code == kCFHostErrorUnknown &&
        [(NSNumber *)error.userInfo[(NSString *)kCFGetAddrInfoFailureKey] intValue] == EAI_NONAME)
    {
        // Reconnect after a long delay and reduce opportunities for a crash to happen (see https://github.com/square/SocketRocket/issues/332)
        [self reopenWebSocket:webSocket afterDelay:30];
    } else {
        [self reopenWebSocket:webSocket afterDelay:1];
    }
}
```

**Comment by groue** [comment]
No news?

**Comment by nlutsenko** [comment]
@groue, looks like this is fixed in #169 and is available on the latest release!

**Comment by groue** [comment]
Thank you @nlutsenko!
